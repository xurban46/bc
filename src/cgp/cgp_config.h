/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

// hard configuration - affects compilation (array sizes etc.)
#define CGP_COLS 8
#define CGP_ROWS 4
#define CGP_INPUTS 9
#define CGP_OUTPUTS 1
#define CGP_LBACK 1
