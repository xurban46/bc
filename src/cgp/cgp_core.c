/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#include <stdlib.h>
#include <string.h>
#include <assert.h>

#include "cgp_core.h"
#include "../random.h"


typedef struct {
    int size;
    int *values;
} int_array;


static int_array _allowed_gene_vals[CGP_COLS];
static int _mutation_rate;
static ga_fitness_func_t _fitness_func;

#ifdef CGP_LIMIT_FUNCS
    static int _allowed_functions_list[] = {
        c255,
        identity,
        add,
        add_sat,
        inversion,
        max,
        min
    };

    static int_array _allowed_functions = {
        .size = 7,
        .values = _allowed_functions_list
    };
#endif


#ifdef TEST_RANDOMIZE
    #define TEST_RANDOMIZE_PRINTF(...) printf(__VA_ARGS__)
#else
    #define TEST_RANDOMIZE_PRINTF(...) /* printf(__VA_ARGS__) */
#endif



/**
 * Initialize CGP internals
 */
void cgp_init(int mutation_rate, ga_fitness_func_t fitness_func)
{
    _mutation_rate = mutation_rate;
    _fitness_func = fitness_func;

    // calculate allowed values of node inputs in each column
    for (int x = 0; x < CGP_COLS; x++) {
        // range of outputs which can be connected to node in i-th column
        // maximum is actually by 1 larger than maximum allowed value
        // (so there is `val < maximum` in the for loop below instead of `<=`)
        int minimum = CGP_ROWS * (x - CGP_LBACK) + CGP_INPUTS;
        if (minimum < CGP_INPUTS) minimum = CGP_INPUTS;
        int maximum = CGP_ROWS * x + CGP_INPUTS;

        int size = CGP_INPUTS + maximum - minimum;
        _allowed_gene_vals[x].size = size;
        _allowed_gene_vals[x].values = (int*) malloc(sizeof(int) * size);

        int key = 0;
        // primary inputs
        for (int val = 0; val < CGP_INPUTS; val++, key++) {
            _allowed_gene_vals[x].values[key] = val;
        }

        // nodes to the left
        for (int val = minimum; val < maximum; val++, key++) {
            _allowed_gene_vals[x].values[key] = val;
        }
    }


#ifdef TEST_INIT
    for (int x = 0; x < CGP_COLS; x++) {
        printf ("x = %d: ", x);
        for (int y = 0; y < _allowed_gene_vals[x].size; y++) {
            if (y > 0) printf(", ");
            printf ("%d", _allowed_gene_vals[x].values[y]);
        }
        printf("\n");
    }
#endif
}


/**
 * Deinitialize CGP internals
 */
void cgp_deinit()
{
    for (int x = 0; x < CGP_COLS; x++) {
        free(_allowed_gene_vals[x].values);
    }
}


/**
 * Create a new cgp population with given size
 * @param  mutation rate (in number of genes)
 * @param  population size
 * @return
 */
ga_pop_t cgp_init_pop(int pop_size)
{
    /* prepare methods vector */
    ga_func_vect_t methods = {
        .alloc_genome = cgp_alloc_genome,
        .free_genome = cgp_free_genome,
        .init_genome = cgp_randomize_genome,

        .fitness = _fitness_func,
        .offspring = cgp_offspring,
    };

    /* initialize GA */
    ga_pop_t pop = ga_create_pop(pop_size, CGP_PROBLEM_TYPE, methods);
    return pop;
}


/* chromosome *****************************************************************/


/**
 * Allocates memory for new CGP genome
 * @param chromosome
 */
void* cgp_alloc_genome()
{
    return malloc(sizeof(struct cgp_genome));
}


/**
 * Initializes CGP genome to random values
 * @param chromosome
 */
int cgp_randomize_genome(ga_chr_t chromosome)
{
    cgp_genome_t genome = (cgp_genome_t) chromosome->genome;

    for (int i = 0; i < CGP_CHR_LENGTH; i++) {
        cgp_randomize_gene(genome, i);
    }

    cgp_find_active_blocks(chromosome);
    chromosome->has_fitness = false;

    return 0;
}


/**
 * Deinitialize CGP genome
 * @param  chromosome
 * @return
 */
void cgp_free_genome(void *genome)
{
    free(genome);
}


/* mutation *******************************************************************/


/**
 * Replace gene on given locus with random alele
 * @param chr
 * @param gene
 * @return whether active node was changed or not (phenotype has changed)
 */
bool cgp_randomize_gene(cgp_genome_t genome, int gene)
{
    if (gene >= CGP_CHR_LENGTH)
        return false;

    if (gene < CGP_CHR_OUTPUTS_INDEX) {
        // mutating node input or function
        int node_index = gene / 3;
        int gene_index = gene % 3;
        int col = cgp_node_col(node_index);

        TEST_RANDOMIZE_PRINTF("gene %u, node %u, gene %u, col %u\n", gene, node_index, gene_index, col);

        if (gene_index == CGP_FUNC_INPUTS) {
            // mutating function
            #ifdef CGP_LIMIT_FUNCS
                genome->nodes[node_index].function = (cgp_func_t) rand_schoice(_allowed_functions.size, _allowed_functions.values);
            #else
                genome->nodes[node_index].function = (cgp_func_t) rand_range(0, CGP_FUNC_COUNT - 1);
            #endif
            TEST_RANDOMIZE_PRINTF("func 0 - %u\n", CGP_FUNC_COUNT - 1);
            return genome->nodes[node_index].is_active;

        } else {
            // mutating input
            genome->nodes[node_index].inputs[gene_index] = rand_schoice(_allowed_gene_vals[col].size, _allowed_gene_vals[col].values);
            TEST_RANDOMIZE_PRINTF("input choice from %u\n", _allowed_gene_vals[col].size);
            return genome->nodes[node_index].is_active;
        }

    } else {
        // mutating primary output connection
        int index = gene - CGP_CHR_OUTPUTS_INDEX;
        genome->outputs[index] = rand_range(CGP_INPUTS, CGP_INPUTS + CGP_NODES - 1);
        TEST_RANDOMIZE_PRINTF("out %u - %u\n", CGP_INPUTS, CGP_INPUTS + CGP_NODES - 1);
        return true;
    }
}


/**
 * Mutate given chromosome
 * @param chr
 */
void cgp_mutate_chr(ga_chr_t chromosome)
{
    assert(_mutation_rate <= CGP_CHR_LENGTH);
    cgp_genome_t genome = (cgp_genome_t) chromosome->genome;

    int genes_to_change = rand_range(0, _mutation_rate);
    for (int i = 0; i < genes_to_change; i++) {
        int gene = rand_range(0, CGP_CHR_LENGTH - 1);
        cgp_randomize_gene(genome, gene);
    }

    cgp_find_active_blocks(chromosome);
    chromosome->has_fitness = false;
}


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_copy_genome(void *_dst, void *_src)
{
    cgp_genome_t dst = (cgp_genome_t) _dst;
    cgp_genome_t src = (cgp_genome_t) _src;

    memcpy(dst->nodes, src->nodes, sizeof(cgp_node_t) * CGP_NODES);
    memcpy(dst->outputs, src->outputs, sizeof(int) * CGP_OUTPUTS);
}


#define SWAP(A, B) (((A & 0x0F) << 4) | ((B & 0x0F)))
#define ADD_SAT(A, B) ((A > 0xFF - B) ? 0xFF : A + B)
#define MAX(A, B) ((A > B) ? A : B)
#define MIN(A, B) ((A < B) ? A : B)


/* evaluation *****************************************************************/


/**
 * Calculate output of given chromosome and inputs
 * @param chr
 */
void cgp_get_output(ga_chr_t chromosome, cgp_value_t *inputs, cgp_value_t *outputs)
{
    cgp_genome_t genome = (cgp_genome_t) chromosome->genome;
    cgp_value_t inner_outputs[CGP_INPUTS + CGP_NODES];

    // copy primary inputs to working array
    memcpy(inner_outputs, inputs, sizeof(cgp_value_t) * CGP_INPUTS);

    for (int i = 0; i < CGP_NODES; i++) {
        cgp_node_t *n = &(genome->nodes[i]);

        // skip inactive blocks
        if (!n->is_active) continue;

        cgp_value_t A = inner_outputs[n->inputs[0]];
        cgp_value_t B = inner_outputs[n->inputs[1]];
        cgp_value_t Y;

        switch (n->function) {
            case c255:          Y = 255;            break;
            case identity:      Y = A;              break;
            case inversion:     Y = 255 - A;        break;
            case b_or:          Y = A | B;          break;
            case b_not1or2:     Y = ~A | B;         break;
            case b_and:         Y = A & B;          break;
            case b_nand:        Y = ~(A & B);       break;
            case b_xor:         Y = A ^ B;          break;
            case rshift1:       Y = A >> 1;         break;
            case rshift2:       Y = A >> 2;         break;
            case swap:          Y = SWAP(A, B);     break;
            case add:           Y = A + B;          break;
            case add_sat:       Y = ADD_SAT(A, B);  break;
            case avg:           Y = (A + B) >> 1;   break;
            case max:           Y = MAX(A, B);      break;
            case min:           Y = MIN(A, B);      break;
            default:            abort();
        }

        inner_outputs[CGP_INPUTS + i] = Y;
    }

    for (int i = 0; i < CGP_OUTPUTS; i++) {
        outputs[i] = inner_outputs[genome->outputs[i]];
    }

#ifdef TEST_EVAL
    for (int i = 0; i < CGP_INPUTS + CGP_NODES; i++) {
        printf("%c: %u = %u\n", (i < CGP_INPUTS? 'I' : 'N'), i, inner_outputs[i]);
    }
    for (int i = 0; i < CGP_OUTPUTS; i++) {
        printf("O: %u = %u\n", i, outputs[i]);
    }
#endif
}


/**
 * Finds which blocks are active.
 * @param chromosome
 * @param active
 */
void cgp_find_active_blocks(ga_chr_t chromosome)
{
    cgp_genome_t genome = (cgp_genome_t) chromosome->genome;

    // first mark all nodes as inactive
    for (int i = CGP_NODES - 1; i >= 0; i--) {
        genome->nodes[i].is_active = false;
    }

    // mark inputs of primary outputs as active
    for (int i = 0; i < CGP_OUTPUTS; i++) {
        int index = genome->outputs[i] - CGP_INPUTS;
        // index may be negative (primary input), so do the check...
        if (index >= 0) {
            genome->nodes[index].is_active = true;
        }
    }

    // then walk nodes backwards and mark inputs of active nodes
    // as active nodes
    for (int i = CGP_NODES - 1; i >= 0; i--) {
        if (!genome->nodes[i].is_active) continue;
        cgp_node_t *n = &(genome->nodes[i]);

        for (int k = 0; k < CGP_FUNC_INPUTS; k++) {
            int index = n->inputs[k] - CGP_INPUTS;
            // index may be negative (primary input), so do the check...
            if (index >= 0) {
                genome->nodes[index].is_active = true;
            }
        }
    }
}



/* population *****************************************************************/


/**
 * Create new generation
 * @param pop
 * @param mutation_rate
 */
void cgp_offspring(ga_pop_t pop)
{
    ga_chr_t parent = pop->best_chromosome;

    #pragma omp parallel for
    for (int i = 0; i < pop->size; i++) {
        ga_chr_t chr = pop->chromosomes[i];
        if (chr == parent) continue;
        ga_copy_chr(chr, parent, cgp_copy_genome);
        cgp_mutate_chr(chr);
    }
}
