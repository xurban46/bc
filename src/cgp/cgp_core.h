/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include "../ga.h"
#include "cgp_config.h"


#define CGP_FUNC_INPUTS 2
#define CGP_NODES (CGP_COLS * CGP_ROWS)
#define CGP_CHR_OUTPUTS_INDEX ((CGP_FUNC_INPUTS + 1) * CGP_NODES)
#define CGP_CHR_LENGTH (CGP_CHR_OUTPUTS_INDEX + CGP_OUTPUTS)

static const ga_problem_type_t CGP_PROBLEM_TYPE = maximize;

typedef unsigned char cgp_value_t;

#define CGP_FUNC_COUNT 16
typedef enum
{
    c255 = 0,   // 255
    identity,   // a
    inversion,  // 255 - a
    b_or,       // a or b
    b_not1or2,  // (not a) or b
    b_and,      // a and b
    b_nand,     // not (a and b)
    b_xor,      // a xor b
    rshift1,    // a >> 1
    rshift2,    // a >> 2
    swap,       // a <-> b
    add,        // a + b
    add_sat,    // a +S b
    avg,        // (a + b) >> 1
    max,        // max(a, b)
    min,        // min(a, b)
} cgp_func_t;


/**
 * One CGP node (function block)
 */
typedef struct {
    int inputs[CGP_FUNC_INPUTS];
    cgp_func_t function;
    bool is_active;
} cgp_node_t;


/**
 * Chromosome
 */
struct cgp_genome {
    cgp_node_t nodes[CGP_COLS * CGP_ROWS];
    int outputs[CGP_OUTPUTS];
};
typedef struct cgp_genome* cgp_genome_t;


/**
 * Initialize CGP internals
 */
void cgp_init(int mutation_rate, ga_fitness_func_t fitness_func);


/**
 * Deinitialize CGP internals
 */
void cgp_deinit();


/**
 * Create a new CGP population with given size
 * @param  mutation rate (in number of genes)
 * @return
 */
ga_pop_t cgp_init_pop(int pop_size);


/**
 * Allocates memory for new CGP genome
 * @return pointer to newly allocated genome
 */
void* cgp_alloc_genome();


/**
 * Initializes CGP genome to random values
 * @param chromosome
 * @return 0 on success, other value on error
 */
int cgp_randomize_genome(ga_chr_t chromosome);


/**
 * Deinitialize CGP genome
 * @param  genome
 * @return
 */
void cgp_free_genome(void *genome);


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_copy_genome(void *_dst, void *_src);


/**
 * Replace gene on given locus with random alele
 * @param chr
 * @param gene
 * @return whether active node was changed or not (phenotype has changed)
 */
bool cgp_randomize_gene(cgp_genome_t genome, int gene);


/**
 * Mutate given chromosome
 * @param chr
 */
void cgp_mutate_chr(ga_chr_t chromosome);


/**
 * Replace chromosome genes with genes from other chromomose
 * @param  chr
 * @param  replacement
 * @return
 */
void cgp_replace_chr(ga_chr_t chromosome, ga_chr_t replacement);


/**
 * Calculate output of given chromosome and inputs
 * @param chr
 */
void cgp_get_output(ga_chr_t chromosome, cgp_value_t *inputs, cgp_value_t *outputs);


/**
 * Create new generation
 * @param pop
 * @param mutation_rate
 */
void cgp_offspring(ga_pop_t pop);


/**
 * Returns index of node in given column and row
 * @param  col
 * @param  row
 * @return
 */
static inline int cgp_node_index(int col, int row)
{
    return CGP_ROWS * col + row;
}


/**
 * Returns node column based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_col(int index)
{
    return index / CGP_ROWS;
}


/**
 * Returns node row based on its index
 * @param  index
 * @return
 */
static inline int cgp_node_row(int index)
{
    return index % CGP_ROWS;
}


/**
 * Finds which blocks are active.
 * @param chromosome
 * @param active
 */
void cgp_find_active_blocks(ga_chr_t chromosome);
