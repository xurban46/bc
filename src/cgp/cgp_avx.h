/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include <immintrin.h>

#include "cgp_core.h"


typedef __m256i __m256i_aligned __attribute__ ((aligned (32)));


/**
 * Calculate output of given chromosome and inputs using AVX instructions
 * @param chr
 * @param inputs
 * @param outputs
 */
void cgp_get_output_avx(ga_chr_t chromosome, __m256i_aligned inputs[CGP_INPUTS], __m256i_aligned outputs[CGP_OUTPUTS]);
