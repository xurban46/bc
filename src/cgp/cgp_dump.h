/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include <stdbool.h>
#include <stdio.h>

#include "cgp_core.h"
#include "cgp_config.h"


/**
 * Dump formats
 */
typedef enum {
    asciiart,
    asciiart_active,
    compat,
    readable,
} cgp_dump_format;



/**
 * Dumps chromosome to given file pointer in given format
 * @param fp
 */
void cgp_dump_chr(ga_chr_t chr, FILE *fp, cgp_dump_format fmt);


/**
 * Dumps chromosome to given file in CGP-viewer compatible format
 * @param chr
 * @param fp
 */
void cgp_dump_chr_compat(ga_chr_t chr, FILE *fp);


/**
 * Dumps chromosome to given file in more human-friendly fashion
 * @param chr
 * @param fp
 */
void cgp_dump_chr_readable(ga_chr_t chr, FILE *fp);


/**
 * Dumps chromosome to given file as an ASCII-art, which looks like this:
 *
 *      .------------------------------------.
 *      |      .----.            .----.      |
 * [ 0]>| [ 1]>|    |>[ 4]  [ 0]>|    |>[ 6] |>[ 6]
 * [ 1]>| [ 3]>|  a |       [ 4]>| xor|      |
 * [ 2]>|      '----'            '----'      |
 * [ 3]>|      .----.            .----.      |
 *      | [ 3]>|    |>[ 5]  [ 4]>|    |>[ 7] |
 *      | [ 0]>|a>>1|       [ 5]>|nand|      |
 *      |      '----'            '----'      |
 *      '------------------------------------'
 *
 * @param chr
 * @param fp
 * @param only_active_blocks Do not render inactive blocks at all
*/
void cgp_dump_chr_asciiart(ga_chr_t chr, FILE *fp, bool only_active_blocks);


static inline int cgp_dump_chr_asciiart_width() {
    return 5 + CGP_COLS * 18 + 5;
}


static inline int cgp_dump_chr_asciiart_height() {
    // header + border + nodes + border
    return 5 + 1 + (CGP_ROWS * 4) + 1;
}


/**
 * Dumps whole population to given file pointer with chromosomes in
 * CGP-viewer compatible format
 * @param pop
 * @param fp
 */
void cgp_dump_pop_compat(ga_pop_t pop, FILE *fp);
