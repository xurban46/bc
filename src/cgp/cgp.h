/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#pragma once

#include "cgp_core.h"
#include "cgp_dump.h"
#include "cgp_load.h"
