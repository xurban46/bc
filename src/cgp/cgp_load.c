/*
 * Colearning in Coevolutionary Algorithms
 * Bc. Michal Wiglasz <xwigla00@stud.fit.vutbr.cz>
 *
 * Master Thesis
 * 2014/2015
 *
 * Supervisor: Ing. Michaela Šikulová <isikulova@fit.vutbr.cz>
 *
 * Faculty of Information Technologies
 * Brno University of Technology
 * http://www.fit.vutbr.cz/
 *
 * Started on 28/07/2014.
 *      _       _
 *   __(.)=   =(.)__
 *   \___)     (___/
 */


#include <stdlib.h>
#include <stdarg.h>

#include "cgp.h"


/**
 * Loads chromosome from given file stored in CGP-viewer compatible format
 * @param chr
 * @param fp
 * @return 0 on success, -1 on file format error, -2 on incompatible CGP config
 */
int cgp_load_chr_compat(ga_chr_t chr, FILE *fp)
{
    int inputs, outputs, cols, rows, func_inputs, func_outputs, func_count;
    int count;
    cgp_genome_t genome = (cgp_genome_t) chr->genome;

    count = fscanf(fp, "{%u, %u, %u, %u, %u, %u, %u}", &inputs, &outputs,
        &cols, &rows, &func_inputs, &func_outputs, &func_count);


    if (count != 7) return -1;
    if (inputs != CGP_INPUTS) return -2;
    if (outputs != CGP_OUTPUTS) return -2;
    if (cols != CGP_COLS) return -2;
    if (rows != CGP_ROWS) return -2;
    if (func_inputs != CGP_FUNC_INPUTS) return -2;
    if (func_outputs != 1) return -2;
    if (func_count != CGP_FUNC_COUNT) return -2;

    // nodes


    for (int i = 0; i < CGP_NODES; i++) {
        cgp_node_t *n = &genome->nodes[i];

        int nodeid;
        count = fscanf(fp, "([%u] %u, %u, %u)",
            &nodeid, &n->inputs[0], &n->inputs[1], &n->function);
        if (count != 4) return -1;
        if (nodeid != CGP_INPUTS + i) return -1;
    }

    // primary outputs


    fscanf(fp, "(");
    for (int i = 0; i < CGP_OUTPUTS; i++) {
        if (i > 0) fscanf(fp, ",");
        count = fscanf(fp, "%u", &genome->outputs[i]);
        if (count != 1) return -1;
    }
    fscanf(fp, ")\n");

    cgp_find_active_blocks(chr);

    return 0;
}
