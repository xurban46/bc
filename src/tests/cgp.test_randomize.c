/**
 * Tests CGP initialization.
 * Compile with -DTEST_RANDOMIZE
 */


#include <stdlib.h>

#include "../cgp.h"


int main(int argc, char const *argv[])
{
    cgp_init(0, NULL);

    cgp_genome_t genome = (cgp_genome_t) malloc(sizeof(struct cgp_genome));
    struct ga_chr chr = {
        .genome = genome
    };

    cgp_randomize_genome(&chr);

    free(genome);
    cgp_deinit();
}
