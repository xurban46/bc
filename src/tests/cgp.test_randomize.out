gene 0, node 0, gene 0, col 0
input choice from 9
gene 1, node 0, gene 1, col 0
input choice from 9
gene 2, node 0, gene 2, col 0
func 0 - 15
gene 3, node 1, gene 0, col 0
input choice from 9
gene 4, node 1, gene 1, col 0
input choice from 9
gene 5, node 1, gene 2, col 0
func 0 - 15
gene 6, node 2, gene 0, col 0
input choice from 9
gene 7, node 2, gene 1, col 0
input choice from 9
gene 8, node 2, gene 2, col 0
func 0 - 15
gene 9, node 3, gene 0, col 0
input choice from 9
gene 10, node 3, gene 1, col 0
input choice from 9
gene 11, node 3, gene 2, col 0
func 0 - 15
gene 12, node 4, gene 0, col 1
input choice from 13
gene 13, node 4, gene 1, col 1
input choice from 13
gene 14, node 4, gene 2, col 1
func 0 - 15
gene 15, node 5, gene 0, col 1
input choice from 13
gene 16, node 5, gene 1, col 1
input choice from 13
gene 17, node 5, gene 2, col 1
func 0 - 15
gene 18, node 6, gene 0, col 1
input choice from 13
gene 19, node 6, gene 1, col 1
input choice from 13
gene 20, node 6, gene 2, col 1
func 0 - 15
gene 21, node 7, gene 0, col 1
input choice from 13
gene 22, node 7, gene 1, col 1
input choice from 13
gene 23, node 7, gene 2, col 1
func 0 - 15
gene 24, node 8, gene 0, col 2
input choice from 13
gene 25, node 8, gene 1, col 2
input choice from 13
gene 26, node 8, gene 2, col 2
func 0 - 15
gene 27, node 9, gene 0, col 2
input choice from 13
gene 28, node 9, gene 1, col 2
input choice from 13
gene 29, node 9, gene 2, col 2
func 0 - 15
gene 30, node 10, gene 0, col 2
input choice from 13
gene 31, node 10, gene 1, col 2
input choice from 13
gene 32, node 10, gene 2, col 2
func 0 - 15
gene 33, node 11, gene 0, col 2
input choice from 13
gene 34, node 11, gene 1, col 2
input choice from 13
gene 35, node 11, gene 2, col 2
func 0 - 15
gene 36, node 12, gene 0, col 3
input choice from 13
gene 37, node 12, gene 1, col 3
input choice from 13
gene 38, node 12, gene 2, col 3
func 0 - 15
gene 39, node 13, gene 0, col 3
input choice from 13
gene 40, node 13, gene 1, col 3
input choice from 13
gene 41, node 13, gene 2, col 3
func 0 - 15
gene 42, node 14, gene 0, col 3
input choice from 13
gene 43, node 14, gene 1, col 3
input choice from 13
gene 44, node 14, gene 2, col 3
func 0 - 15
gene 45, node 15, gene 0, col 3
input choice from 13
gene 46, node 15, gene 1, col 3
input choice from 13
gene 47, node 15, gene 2, col 3
func 0 - 15
gene 48, node 16, gene 0, col 4
input choice from 13
gene 49, node 16, gene 1, col 4
input choice from 13
gene 50, node 16, gene 2, col 4
func 0 - 15
gene 51, node 17, gene 0, col 4
input choice from 13
gene 52, node 17, gene 1, col 4
input choice from 13
gene 53, node 17, gene 2, col 4
func 0 - 15
gene 54, node 18, gene 0, col 4
input choice from 13
gene 55, node 18, gene 1, col 4
input choice from 13
gene 56, node 18, gene 2, col 4
func 0 - 15
gene 57, node 19, gene 0, col 4
input choice from 13
gene 58, node 19, gene 1, col 4
input choice from 13
gene 59, node 19, gene 2, col 4
func 0 - 15
gene 60, node 20, gene 0, col 5
input choice from 13
gene 61, node 20, gene 1, col 5
input choice from 13
gene 62, node 20, gene 2, col 5
func 0 - 15
gene 63, node 21, gene 0, col 5
input choice from 13
gene 64, node 21, gene 1, col 5
input choice from 13
gene 65, node 21, gene 2, col 5
func 0 - 15
gene 66, node 22, gene 0, col 5
input choice from 13
gene 67, node 22, gene 1, col 5
input choice from 13
gene 68, node 22, gene 2, col 5
func 0 - 15
gene 69, node 23, gene 0, col 5
input choice from 13
gene 70, node 23, gene 1, col 5
input choice from 13
gene 71, node 23, gene 2, col 5
func 0 - 15
gene 72, node 24, gene 0, col 6
input choice from 13
gene 73, node 24, gene 1, col 6
input choice from 13
gene 74, node 24, gene 2, col 6
func 0 - 15
gene 75, node 25, gene 0, col 6
input choice from 13
gene 76, node 25, gene 1, col 6
input choice from 13
gene 77, node 25, gene 2, col 6
func 0 - 15
gene 78, node 26, gene 0, col 6
input choice from 13
gene 79, node 26, gene 1, col 6
input choice from 13
gene 80, node 26, gene 2, col 6
func 0 - 15
gene 81, node 27, gene 0, col 6
input choice from 13
gene 82, node 27, gene 1, col 6
input choice from 13
gene 83, node 27, gene 2, col 6
func 0 - 15
gene 84, node 28, gene 0, col 7
input choice from 13
gene 85, node 28, gene 1, col 7
input choice from 13
gene 86, node 28, gene 2, col 7
func 0 - 15
gene 87, node 29, gene 0, col 7
input choice from 13
gene 88, node 29, gene 1, col 7
input choice from 13
gene 89, node 29, gene 2, col 7
func 0 - 15
gene 90, node 30, gene 0, col 7
input choice from 13
gene 91, node 30, gene 1, col 7
input choice from 13
gene 92, node 30, gene 2, col 7
func 0 - 15
gene 93, node 31, gene 0, col 7
input choice from 13
gene 94, node 31, gene 1, col 7
input choice from 13
gene 95, node 31, gene 2, col 7
func 0 - 15
out 9 - 40
